Element Reordering and Field Group Averaging via MPI

Udayaputta99 opened the pull request:
Scalable MPI-based:
1. Efficiently reorders elements and groups them into batches.
2. Method to compute averages (avg_<key>) for all fields defined in field_keys.

Udayaputta99 commented:
Hi! I merged upstream/main into my fork, but I’m seeing some merge conflicts in files I haven’t worked on. Could you help clarify which changes should be kept?

adperezm commented on pysemtools/io/reorder_data.py:
```diff
@@ -0,0 +1,316 @@
+
+import os
+import sys
+sys.path.append("/home/hpc/iwst/iwst115h/forked_git/pySEMTools/pysemtools")
```
Remove this.

Udayaputta99 replied:
I have removed the custom path references as requested

adperezm commented on pysemtools/io/reorder_data.py:
```diff
@@ -0,0 +1,316 @@
+
+def generate_data(fname_3d: str, fname_out: str):
+
+    msh_3d = Mesh(comm, create_connectivity=True)
```
I would not create connectivity here unless really needed. This makes it much slower.

adperezm commented on examples/10-mapping-grouping/elem_group.py:
```diff
@@ -0,0 +1,528 @@
+    factor = 10.0 ** decimals
+    return np.floor(trial * factor) / factor
+
+def extracted_fields(fld_3d_r):
```
I do not get this function. Why not using a FildRegistry class for fld_3d_r? It seems to me as if that is exactly what this function is doing, or?

adperezm commented on examples/10-mapping-grouping/elem_group.py:
```diff
@@ -0,0 +1,528 @@
+    fname_3d = "/home/woody/iwst/iwst115h/Work/Checkpoint_files/field0.f00000"
+    
+    # fname_out = "/home/woody/iwst/iwst115h/Work/writefiles/unstruct_new/field_out_0.f00024"
+    fname_out = "/home/woody/iwst/iwst115h/Work/writefiles/field/fieldout0.f00000"
```
If possible, use data in the example repository. If this is an example, it can't be run based on these paths.

adperezm replied:
Otherwise, you can also upload some data to the example repository (if it is a small case)

Udayaputta99 replied:
Yeah, I have specified a relative path. I can't add example files since they are very large.

adperezm reviewed (changes requested):
Hi,

I think there should be some changes. Furthermore, can you add docstrings to the functions? I am not exactly sure what they do.

Lastly, it seems that you are kinda using all_to_all operations. Why not use the communication object that we have available and also the mesh partitioning object we have?

Could you explain to me a bit what is the intent of these things?

Udayaputta99 commented:
1. I have added docstrings to the functions. Regarding all-to-all communication, I explicitly decided not to use it because the data being exchanged is very large. Instead, I ensured that each rank only receives the specific data it requires from other ranks, optimizing memory usage.
 
2. Additionally, I have implemented the suggested mesh partitioning using "conditions" to improve data distribution.

3.  **Could you explain to me a bit what is the intent of these things?**
In a nutshell, I am averaging the fields in the data files across repeating patterns. By doing this, we eliminate redundant data and retain only the repeating statistical representation, which significantly reduces storage.

I've updated the requested changes. Let me know if anything else needs adjustment.
